lib/bolt_server/transport_app.rb

```diff
@@ -675,6 +681,56 @@ def make_winrm_target(target_hash)
       end
     end
 
+    # Returns the base64 encoded tar archive of plugin code that is needed to calculate
+    # custom facts
+    #
+    # @param versioned_project [String] the versioned_project to build the plugin tarball from
+    post '/project_facts_plugin_tarball' do
+      raise BoltServer::RequestError, "'versioned_project' is a required argument" if params['versioned_project'].nil?
+      content_type :json
+      in_bolt_project(params['versioned_project']) do |context|
+        # Inspired by Bolt::Applicator.build_plugin_tarball
+        start_time = Time.now
+        sio = StringIO.new
+
+        begin
+          output = Minitar::Output.new(Zlib::GzipWriter.new(sio))
+
+          # Bolt also sets this to user modulepath so do it here too
+          # for consistency
+          plugin_modulepath = context[:pal].user_modulepath
+          Puppet.lookup(:current_environment).override_with(modulepath: plugin_modulepath).modules.each do |mod|
```

Comment:
Can we release the PAL lock quicker than this (maybe just look up all the files to pack, then pack them after we are out of the project context?

Reply:
Sure, updated.